broker/cloud_functions/filter_purity/main.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+# -*- coding: UTF-8 -*-
+
+"""Filter alerts for purity."""
+
+import base64
+from google.cloud import logging
+import os
+
+# Add these to requirements.txt but check to make sure that the format of adding is correct
+from broker_utils import data_utils, gcp_utils, schema_maps 
+
+
+
+PROJECT_ID = os.getenv("GCP_PROJECT")
+TESTID = os.getenv("TESTID") # I am guessing this returns True?
+SURVEY = os.getenv("SURVEY") # I am guessing this return ztf
+
+# connect to the logger
+logging_client = logging.Client()
+log_name = "filter-purity"  # same log for all broker instances
+logger = logging_client.logger(log_name)
+
+# GCP resources used in this module
+bq_dataset = f"{SURVEY}_alerts"
+
+# Changed the name stub of ps_topic from exgalac_trans_cs to alerts_pure
+ps_topic = f"{SURVEY}-alerts_pure" ## I think that this is for BigQuery Database Storage name stub
+if TESTID != "False":  # attach the testid to the names
+    bq_dataset = f"{bq_dataset}_{TESTID}"
+    ps_topic = f"{ps_topic}-{TESTID}"
+
+# Changed the name stub of the BigQuery table from SuperNNova to classifications
+bq_table = f"{bq_dataset}.classifications"
+
+schema_map = schema_maps.load_schema_map(SURVEY, TESTID)
+
+
+def is_pure(alert, schema_map):
+    """Adapted from: https://zwickytransientfacility.github.io/ztf-avro-alert/filtering.html
+
+    Quoted from the source:
+
+    ZTF alert streams contain an nearly entirely unfiltered stream of all
+    5-sigma (only the most obvious artefacts are rejected). Depending on your
+    science case, you may wish to improve the purity of your sample by filtering
+    the data on the included attributes.
+
+    Based on tests done at IPAC (F. Masci, priv. comm), the following filter
+    delivers a relatively pure sample.
+    """
+    source = alert[schema_map['source']]
+
+    rb = (source['rb'] >= 0.65)  # RealBogus score
+
+    if schema_map['SURVEY'] == 'decat':
+        is_pure = rb
+
+    elif schema_map['SURVEY'] == 'ztf':
+        nbad = (source['nbad'] == 0)  # num bad pixels
+        fwhm = (source['fwhm'] <= 5)  # Full Width Half Max, SExtractor [pixels]
+        elong = (source['elong'] <= 1.2)  # major / minor axis, SExtractor
+        magdiff = (abs(source['magdiff']) <= 0.1)  # aperture - psf [mag]
+        is_pure = (rb and nbad and fwhm and elong and magdiff)
+
+    return is_pure
```

Comment:
Update this function so that it returns a dictionary with 6 key / value pairs: 1 for the final result (`is_pure`) and one for each of the 5 cuts (`rb`, `nbad`, etc.). We want to save all of these results so that later we can look back and know exactly why an alert did / didn't pass this filter.

Make sure you update the `run()` function to accommodate the new output.